### PR TITLE
Multifurcating tree to Bifurcating tree

### PR DESCRIPTION
**Pull request by mortonjt**
This PR is to address https://github.com/biocore/scikit-bio/issues/896

Some of the analyses I've been doing has required strictly bifurcating trees, so this method 
1. Introduces additional internal nodes to ensure that all the internal nodes have exactly 2 children
2. Removes nodes that have exactly one child (if `remove_singles=True`)

The order in which these internal nodes are injected or how the single child nodes are removed are a little arbitrary at the moment.

Todo
- [x] squash commits

### CODE REVIEW

**Comment by Jorge-C** [comment]
There are some minor failures related to whitespace in docstrings I think.

**Comment by wasade** [comment]
This method doesn't handle branch length, possible to add support?

**Comment by mortonjt** [comment]
:+1: 

Perhaps, I could split up the lengths so that the original lengths are preserved.
For instance, if the multifurcated tree and its bifurcating equivalent are below
```
     /---  B
A --|----  C
     \---  D
```

```
     /---  B
A --|
    |     /--- C
     \---X
          \--- D
```
And the lengths of the branches for the first tree are all one-third
Then the lengths of the branches for the second tree are 
```
B = 1/3
X = 1/6
C = 1/6
D = 1/6
```

Is this something that you were considering, or did you have something else in mind @wasade ?

**Comment by gregcaporaso** [comment]
It's probably worth checking how other packages handle this before deciding how we want to. For example, I'd start by looking at [phylip](http://evolution.genetics.washington.edu/phylip.html) and [ete2](http://etetoolkit.org/).

**Comment by wasade** [comment]
@mortonjt, seems plausible but agree with @gregcaporaso. phylip is the de
facto standard, and that site should be pretty well documented.

**Comment by mortonjt** [comment]
ETE has a method that converts multifurcating trees to bifurcating trees: see [`resolve_polytomy()`](https://github.com/jhcepas/ete/blob/master/ete2/coretype/tree.py#L2443)

Phylip doesn't seem to explicitly have this functionality, but it has a program [`retree`](http://evolution.genetics.washington.edu/phylip/doc/retree.html) that can allow the user to manually rearrange their tree

APE also has a method that converts multifurcating trees to bifurcating trees.  see [`multi2di`](http://www.inside-r.org/packages/cran/ape/docs/multi2di)

Couldn't find any such functionality in NetworkX

tl;dr There doesn't seem to be an established standard for converting between multifurcating trees and bifurcating trees

**Comment by gregcaporaso** [comment]
The ete approach seems reasonable to me...

**Comment by gregcaporaso** [comment]
Thanks for looking into this @mortonjt!

**Comment by mortonjt** [comment]
I think there may be an infinite loop that occurs after calling `assign_ids` after `bifurcate` has been applied to larger multi-furcating trees

Not exactly sure what is going on.  After talking to @wasade offline about this, I'm considering pushing another PR to add functionality to check to see if there are any errors in the tree (aka cycles or multiple parents).

File: skbio/tree/_tree.py [hunk 1]
```diff
@@ -2649,6 +2649,54 @@ def compare_tip_distances(self, other, sample=None, dist_f=distance_from_r,
...
+            if n.is_tip():
+                continue
+            if len(n.children) == 1 and remove_singles:
+                i = n.children[0]
+                if i.is_tip():
+                    n.parent.append(i)
+                    n.parent.remove(n)
+                else:
+                    n.extend(i.children)
+                    n.remove(i)
```
Comment (wasade) [inline skbio/tree/_tree.py]: I think the issues are the remove statements. `extend` and `append` call `_adopt` which modifies references

Comment (mortonjt) [inline skbio/tree/_tree.py]: Looks like removing those `remove` calls fixed it.  Thanks!

**Comment by jairideout** [comment]
@mortonjt besides the merge conflict what are the next steps for this PR?

**Comment by mortonjt** [comment]
Probably should add a [DO NOT MERGE] tag on this PR.

So removing those `remove` statements doesn't fix the problem - there are other cases where this fails.
I'm not entirely sure what's going on with the assign_ids() problem.  This occurs in one of my real datasets with 800 some nodes.  Still trying to figure out exactly why this is occuring.

File: skbio/tree/_tree.py [hunk 2]
```diff
@@ -2649,6 +2649,55 @@ def compare_tip_distances(self, other, sample=None, dist_f=distance_from_r,
...
+            elif len(n.children) > 2:
+                stack = n.children
+                curnode = n
+                while len(stack) > 2:
+                    ind = stack.pop()
+                    intermediate = TreeNode()
+                    intermediate.length = insert_length
+                    curnode.children = [ind, intermediate]
```
Comment (wasade) [inline skbio/tree/_tree.py]: these should use `remove`, `append`, and/or `extend` to make sure the references are maintained correctly. that is likely why `assign_ids` is acting funny

Comment (mortonjt) [inline skbio/tree/_tree.py]: If I remove `assign_ids`, this function worked exactly how I need it to.

The question is, is `assign_ids()` strictly necessary?

**Comment by jairideout** [comment]
@mortonjt thanks, I tagged this as "in progress".

**Comment by mortonjt** [comment]
Ok.  The infinite loop problem is __gone__!

The problem was some of the parent pointers weren't being appropriately assigned when producing intermediate nodes.

Soooooo.  I think this is up for a proper review.  If you think that there are any additional test cases that I need to consider.  Thanks!

**Comment by jairideout** [comment]
Great, thanks @mortonjt!

@wasade are you available to review this?

**Comment by wasade** [comment]
Yes, but not for a few days. Can you assign please?

**Comment by mortonjt** [comment]
Just realized that `TreeNode.prune` already removes internal nodes with single children, :santa: 

So that's awesome - the code should now be a bit more clean

**Comment by jairideout** [comment]
@wasade assigned, thanks!

File: skbio/tree/_tree.py [hunk 3]
```diff
@@ -2705,6 +2705,47 @@ def compare_tip_distances(self, other, sample=None, dist_f=distance_from_r,
         return dist_f(self_matrix, other_matrix)
 
     @experimental(as_of="0.4.0")
+    def bifurcate(self, insert_length=None, remove_singles=True):
+        r""" Restructures tree into a bifurcating tree
```
Comment (wasade) [inline skbio/tree/_tree.py]: extra space, possible to remove?

Comment (wasade) [inline skbio/tree/_tree.py]: "Restructures the tree..."

File: skbio/tree/_tree.py [hunk 4]
```diff
@@ -2705,6 +2705,47 @@ def compare_tip_distances(self, other, sample=None, dist_f=distance_from_r,
...
+
+        All nodes that have more than 2 children will
```
Comment (wasade) [inline skbio/tree/_tree.py]: It looks like this paragraph could be refit at 80 chars (as in, the line breaks appear premature

File: skbio/tree/_tree.py [hunk 5]
```diff
@@ -2705,6 +2705,47 @@ def compare_tip_distances(self, other, sample=None, dist_f=distance_from_r,
...
+        have additional intermediate nodes injected to ensure that
```
Comment (wasade) [inline skbio/tree/_tree.py]: injected is unusual here, what about inserted?

File: skbio/tree/_tree.py [hunk 6]
```diff
@@ -2705,6 +2705,47 @@ def compare_tip_distances(self, other, sample=None, dist_f=distance_from_r,
...
+        every node has only 2 children
+        Also, any nodes that have a single child will be collapsed
```
Comment (wasade) [inline skbio/tree/_tree.py]: I think this should be added to a `Notes` section.

File: skbio/tree/_tree.py [hunk 7]
```diff
@@ -2705,6 +2705,47 @@ def compare_tip_distances(self, other, sample=None, dist_f=distance_from_r,
         return dist_f(self_matrix, other_matrix)
 
     @experimental(as_of="0.4.0")
+    def bifurcate(self, insert_length=None, remove_singles=True):
+        r""" Restructures tree into a bifurcating tree
+
+        All nodes that have more than 2 children will
+        have additional intermediate nodes injected to ensure that
+        every node has only 2 children
```
Comment (wasade) [inline skbio/tree/_tree.py]: This is inaccurate in the case of a tree with an odd number of tips. It may be useful to distinguish "node" from "tip" as this method should ensure that _internal_ nodes have at most 2 children, but not that every node has at most 2 children.

File: skbio/tree/_tree.py [hunk 8]
```diff
@@ -2705,6 +2705,47 @@ def compare_tip_distances(self, other, sample=None, dist_f=distance_from_r,
...
+        Also, any nodes that have a single child will be collapsed
+
+        Parameters
+        ----------
+        remove_singles : bool, optional
```
Comment (wasade) [inline skbio/tree/_tree.py]: since all this does is call `prune`, should this be part of this method?

File: skbio/tree/_tree.py [hunk 9]
```diff
@@ -2705,6 +2705,47 @@ def compare_tip_distances(self, other, sample=None, dist_f=distance_from_r,
...
+            Removes all nodes that have only a single child and relink
+            of their childrent to their parent node
+
+        Examples
+        --------
+        >>> from skbio import TreeNode
+        >>> from six import StringIO
+        >>> tree = TreeNode.read(StringIO("((a,b,g,h)c,(d,e)f)root;"))
```
Comment (wasade) [inline skbio/tree/_tree.py]: minor, and non-blocking, but it would improve the clarity of the example to print `ascii_art` prior to calling `bifurcate` as well as printing `ascii_art` after the call.

Comment (mortonjt) [inline skbio/tree/_tree.py]: :+1:

File: skbio/tree/_tree.py [hunk 10]
```diff
@@ -2705,6 +2705,47 @@ def compare_tip_distances(self, other, sample=None, dist_f=distance_from_r,
         return dist_f(self_matrix, other_matrix)
 
     @experimental(as_of="0.4.0")
+    def bifurcate(self, insert_length=None, remove_singles=True):
```
Comment (wasade) [inline skbio/tree/_tree.py]: Open question: should this be generalized to support N > 1 case or would it not be beneficial to coerce to, say, trifurcating?

Comment (mortonjt) [inline skbio/tree/_tree.py]: That is a very doable change.

However, is it useful?

File: skbio/tree/_tree.py [hunk 11]
```diff
@@ -2705,6 +2705,47 @@ def compare_tip_distances(self, other, sample=None, dist_f=distance_from_r,
...
+        >>> print(tree)
+        ((h,(g,(a,b)))c,(d,e)f)root;
+        <BLANKLINE>
+
+        """
+        for n in self.traverse(include_self=True):
+            if len(n.children) > 2:
+                stack = n.children
+                while len(stack) > 2:
+                    ind = stack.pop()
```
Comment (wasade) [inline skbio/tree/_tree.py]: Are you sure this shouldn't be `n.pop`? I'm concerned about the potential for references to get confused here.

File: skbio/tree/_tree.py [hunk 12]
```diff
@@ -2705,6 +2705,47 @@ def compare_tip_distances(self, other, sample=None, dist_f=distance_from_r,
...
+                    intermediate = TreeNode()
+                    intermediate.length = insert_length
+                    intermediate.extend(stack)
+                    intermediate.parent = n
```
Comment (wasade) [inline skbio/tree/_tree.py]: You should not set `parent` explicitly, but instead use either `parent.append` or `parent.extend`.

File: skbio/tree/_tree.py [hunk 13]
```diff
@@ -2705,6 +2705,47 @@ def compare_tip_distances(self, other, sample=None, dist_f=distance_from_r,
...
+                    for k in stack:
+                        n.remove(k)
+                    n.extend([ind, intermediate])
+        if remove_singles:
+            self.prune()
+        self.assign_ids()
```
Comment (wasade) [inline skbio/tree/_tree.py]: Is this necessary?

File: skbio/tree/tests/test_tree.py [hunk 14]
```diff
@@ -784,6 +784,39 @@ def test_levelorder(self):
...
+    def test_bifurcate(self):
+        t1 = TreeNode.read(StringIO(u'(((a,b),c),(d,e));'))
+        t2 = TreeNode.read(StringIO(u'((((((a,b),(c,d)),(e,f))J)Q)K);'))
+        t3 = TreeNode.read(StringIO(u'(((a,b,c),(d)),(e,f));'))
+        t4 = TreeNode.read(StringIO(u'(((a,b,c),(d)),(e,f));'))
+        t5 = TreeNode.read(StringIO(u'(((c,(f)d,(g,h)e)b)a);'))
+        t6 = TreeNode.read(StringIO(u'((((g,h)e,(f)d,c)b)a);'))
+        t7 = TreeNode.read(StringIO(u'((c,((g,h)e,f))b);\n'))
+        t8 = TreeNode.read(StringIO(u'((((g,h)e,c,(f)d)b)a);'))
+        t9 = TreeNode.read(StringIO(u'((((g,h)e,c,(f)d)b)a);'))
```
Comment (wasade) [inline skbio/tree/tests/test_tree.py]: `t8` and `t9` are identical, as are `t3` and `t4`. Could just use `.copy`

Possible to add a test case in which the number of tips in the tree is odd?

File: skbio/tree/_tree.py (hunk 10 above)
Comment (wasade) [inline skbio/tree/_tree.py]: Unsure, but here's the general [case](https://github.com/pycogent/pycogent/blob/master/cogent/core/tree.py#L1417). I confess, I thought we'd ported this code but it doesn't appear to be in skbio...

File: skbio/tree/_tree.py (hunk 8 above)
Comment (mortonjt) [inline skbio/tree/_tree.py]: So that was one thing that I was a little hung up on when putting together this method.  Because I wanted to enable this method to allow for both bifurcating trees (internal nodes have at most 2 children) and strictly bifurcating trees (internal nodes have strictly 2 children)

I'm fine with removing the `remove_singles` parameters.  However, would it become more unintuitive to create strictly bifurcating trees for the average user?

To create a strictly bifurcating tree, it would basically change the following command from
```
tree.bifurcate(remove_singles=True)
``` 
To
```
tree.bifurcate()
tree.prune()
```

In hindsight, I think it is ok to remove this parameter

File: skbio/tree/_tree.py (hunk 13 above)
Comment (mortonjt) [inline skbio/tree/_tree.py]: Probably not.  Consider it removed

**Comment by mortonjt** [comment]
Really appreciate all of the comments @wasade ! Will be incorporating them all shortly

File: skbio/tree/_tree.py (hunk 7 above)
Comment (wasade) [inline skbio/tree/_tree.py]: ...i take this back, for bifurcating, I believe it is always true that every node has 2 children. Higher N I don't think can satisfy that constraint

Comment (mortonjt) [inline skbio/tree/_tree.py]: Alright, so remove the `remove_singles` parameter and make it strictly bifurcating?

I wasn't sure about the precise definition.  The application that I'm developing this for requires strictly bifurcating trees.  I'm honestly not sure how useful having a bifurcating tree (i.e. every intermediate node has at least two nodes) would be.  This was the original intent on including the `remove_singles` parameter

Comment (ebolyen) [inline skbio/tree/_tree.py]: What about the case of an unrooted bifurcating tree? There is a common convention to have exactly 3 children at the root and the rest of the tree is bifurcating.

Comment (wasade) [inline skbio/tree/_tree.py]: Excellent point

Comment (mortonjt) [inline skbio/tree/_tree.py]: :bomb:

Now, this can be done with the current functionality in the PR.
It would be something like this
```
tree.children[0].bifurcate()
tree.children[1].bifurcate()
tree.children[2].bifurcate()
```
I admit, this is hack and not elegant.  What would be a proper alternative to handle this scenario?  Would it be helpful to add an additional parameter (i.e. `unrooted=True`)?

**Comment by mortonjt** [comment]
For some reason, my doctests are all crashing on `skbio.TreeNode`.  Is anyone else having this problem?

```
In [1]: from skbio import TreeNode

In [2]: tree = TreeNode.read(["(((a,b)c,(d,e)f)g,h)i;"])
---------------------------------------------------------------------------
IOSourceError                             Traceback (most recent call last)
<ipython-input-2-a4a457dd8868> in <module>()
----> 1 tree = TreeNode.read(["(((a,b)c,(d,e)f)g,h)i;"])

/Users/mortonjt/.virtualenvs/skbio/lib/python2.7/site-packages/scikit_bio-0.4.0_dev-py2.7-macosx-10.10-intel.egg/skbio/io/registry.pyc in read(cls, file, format, **kwargs)
    622         @classmethod
    623         def read(cls, file, format=None, **kwargs):
--> 624             return registry.read(file, into=cls, format=format, **kwargs)
    625 
    626         imports = registry._import_paths(read_formats)

/Users/mortonjt/.virtualenvs/skbio/lib/python2.7/site-packages/scikit_bio-0.4.0_dev-py2.7-macosx-10.10-intel.egg/skbio/io/registry.pyc in read(self, file, format, into, verify, **kwargs)
    495             return (x for x in itertools.chain([next(gen)], gen))
    496         else:
--> 497             return self._read_ret(file, format, into, verify, kwargs)
    498 
    499     def _read_ret(self, file, fmt, into, verify, kwargs):

/Users/mortonjt/.virtualenvs/skbio/lib/python2.7/site-packages/scikit_bio-0.4.0_dev-py2.7-macosx-10.10-intel.egg/skbio/io/registry.pyc in _read_ret(self, file, fmt, into, verify, kwargs)
    499     def _read_ret(self, file, fmt, into, verify, kwargs):
    500         io_kwargs = self._find_io_kwargs(kwargs)
--> 501         with _resolve_file(file, **io_kwargs) as (file, _, _):
    502             reader, kwargs = self._init_reader(file, fmt, into, verify, kwargs,
    503                                                io_kwargs)

/Users/mortonjt/.virtualenvs/skbio/lib/python2.7/site-packages/contextlib2.pyc in __enter__(self)
     47     def __enter__(self):
     48         try:
---> 49             return next(self.gen)
     50         except StopIteration:
     51             raise RuntimeError("generator didn't yield")

/Users/mortonjt/.virtualenvs/skbio/lib/python2.7/site-packages/scikit_bio-0.4.0_dev-py2.7-macosx-10.10-intel.egg/skbio/io/util.pyc in _resolve_file(file, **kwargs)
    202 @contextmanager
    203 def _resolve_file(file, **kwargs):
--> 204     file, source, is_binary_file = _resolve(file, **kwargs)
    205     try:
    206         yield file, source, is_binary_file

/Users/mortonjt/.virtualenvs/skbio/lib/python2.7/site-packages/scikit_bio-0.4.0_dev-py2.7-macosx-10.10-intel.egg/skbio/io/util.pyc in _resolve(file, mode, encoding, errors, newline, compression, compresslevel)
     56     for source_handler in get_io_sources():
     57         source = source_handler(file, arguments)
---> 58         if mode == 'r' and source.can_read():
     59             newfile = source.get_reader()
     60             break

/Users/mortonjt/.virtualenvs/skbio/lib/python2.7/site-packages/scikit_bio-0.4.0_dev-py2.7-macosx-10.10-intel.egg/skbio/io/_iosources.pyc in can_read(self)
    177                 raise IOSourceError(
    178                     "Could not open source: %r (mode: %r)" %
--> 179                     (self.file, self.options['mode']))
    180         return False
    181 

IOSourceError: Could not open source: ['(((a,b)c,(d,e)f)g,h)i;'] (mode: 'r')

```

**Comment by ebolyen** [comment]
Doctests are run in Py3 now. There's a PR in to make that error better for
Py2.

**Comment by mortonjt** [comment]
ahhh perfect.

File: skbio/tree/_tree.py (hunk 12 above)
Comment (mortonjt) [inline skbio/tree/_tree.py]: Done

File: skbio/tree/tests/test_tree.py (hunk 14 above)
Comment (mortonjt) [inline skbio/tree/tests/test_tree.py]: Done

File: skbio/tree/_tree.py (hunk 11 above)
Comment (mortonjt) [inline skbio/tree/_tree.py]: I'm basically creating a stack for all of the children and reordering the children if there are more than 2 children.  

The references shouldn't change since this is an auxiliary data structure that isn't explicitly touching the tree itself.  As far as I can tell, it isn't a problem considering the tests are behaving as expected.  Do you have suggestions on how to test this?

**Comment by mortonjt** [comment]
This is weird, it is failing in the redundancy analysis and composition tests. @jairideout @ebolyen any idea what is going on?

**Comment by jairideout** [comment]
@mortonjt I just created #1267, the RDA and composition test failures are unrelated to your changes. Note that there is a legitimate `bifurcate` test failure in the build.

**Comment by mortonjt** [comment]
Commits are squashed and tests should be passing.

@wasade , did I address all of your concerns?

**Comment by wasade** [comment]
:+1:

File: skbio/tree/_tree.py [hunk 15]
```diff
@@ -2717,6 +2717,52 @@ def compare_tip_distances(self, other, sample=None, dist_f=distance_from_r,
         return dist_f(self_matrix, other_matrix)
 
     @experimental(as_of="0.4.0")
```
Comment (jairideout) [inline skbio/tree/_tree.py]: Change to `0.4.1-dev`

File: skbio/tree/_tree.py [hunk 16]
```diff
@@ -2717,6 +2717,52 @@ def compare_tip_distances(self, other, sample=None, dist_f=distance_from_r,
...
+    def bifurcate(self, insert_length=None, remove_singles=True):
+        r"""Reorders the tree into a bifurcating tree, so that
```
Comment (jairideout) [inline skbio/tree/_tree.py]: The first line of a docstring must be a single sentence within 79 characters, otherwise it will render incorrectly (note the `r` prefix is unnecessary here)

File: skbio/tree/_tree.py [hunk 17]
```diff
@@ -2717,6 +2717,52 @@ def compare_tip_distances(self, other, sample=None, dist_f=distance_from_r,
...
+        have additional intermediate nodes inserted to ensure that
+        every node has only 2 children.
+
+        Parameters
+        ----------
+        insert_length : int, optional
+            The branch length assigned to all inserted nodes.
+
+        remove_singles : bool, optional
+            Removes all nodes that have only a single child and relink
```
Comment (jairideout) [inline skbio/tree/_tree.py]: I'm having trouble understanding this sentence, is this right?

*Removes all nodes that have only a single child and relinks those nodes' children with their parent nodes.*

File: skbio/tree/_tree.py [hunk 18]
```diff
@@ -2717,6 +2717,52 @@ def compare_tip_distances(self, other, sample=None, dist_f=distance_from_r,
...
+
+        Examples
+        --------
+        >>> from skbio import TreeNode
+        >>> tree = TreeNode.read(["((a,b,g,h)c,(d,e)f)root;"])
+        >>> tree.bifurcate()
+        >>> print(tree)
+        ((h,(g,(a,b)))c,(d,e)f)root;
+        <BLANKLINE>
+
```
Comment (jairideout) [inline skbio/tree/_tree.py]: remove extra newline

File: skbio/tree/_tree.py [hunk 19]
```diff
@@ -2717,6 +2717,52 @@ def compare_tip_distances(self, other, sample=None, dist_f=distance_from_r,
...
+
+        Notes
```
Comment (jairideout) [inline skbio/tree/_tree.py]: move Notes section above Examples

File: skbio/tree/_tree.py [hunk 20]
```diff
@@ -2717,6 +2717,52 @@ def compare_tip_distances(self, other, sample=None, dist_f=distance_from_r,
...
+        -----
+        Any nodes that have a single child will be collapsed.
```
Comment (jairideout) [inline skbio/tree/_tree.py]: Can this sentence be removed since you added `remove_singles`?

File: skbio/tree/_tree.py [hunk 21]
```diff
@@ -2717,6 +2717,52 @@ def compare_tip_distances(self, other, sample=None, dist_f=distance_from_r,
...
+            The branch length assigned to all inserted nodes.
+
+        remove_singles : bool, optional
+            Removes all nodes that have only a single child and relink
+            of their childrent to their parent node.
+
+        Examples
+        --------
+        >>> from skbio import TreeNode
+        >>> tree = TreeNode.read(["((a,b,g,h)c,(d,e)f)root;"])
```
Comment (jairideout) [inline skbio/tree/_tree.py]: Minor, but it might make the example easier to understand if you showed the tree's ASCII art before and after bifurcation.

File: skbio/tree/_tree.py [hunk 22]
```diff
@@ -2717,6 +2717,52 @@ def compare_tip_distances(self, other, sample=None, dist_f=distance_from_r,
...
+        All nodes that have more than 2 children will
+        have additional intermediate nodes inserted to ensure that
+        every node has only 2 children.
+
+        Parameters
+        ----------
+        insert_length : int, optional
+            The branch length assigned to all inserted nodes.
+
+        remove_singles : bool, optional
```
Comment (jairideout) [inline skbio/tree/_tree.py]: All this parameter does is call `prune()`. Would users expect that singles are removed by default? If not, I think we can remove this parameter and add a mention in the Notes section about calling `prune()` if the user wants to remove singles.

**Comment by jairideout** [comment]
@mortonjt I'm done reviewing, can you also add a changelog mention under 0.4.1-dev "Features" heading?

**Comment by jairideout** [comment]
Note that we'll need these changes by tomorrow morning to include in 0.4.2.

File: skbio/tree/_tree.py (hunk 19 above)
Comment (mortonjt) [inline skbio/tree/_tree.py]: Since we are removing remove_singles, the Notes section is not necessary

File: skbio/tree/_tree.py (hunk 20 above)
Comment (mortonjt) [inline skbio/tree/_tree.py]: Right

File: skbio/tree/_tree.py (hunk 21 above)
Comment (mortonjt) [inline skbio/tree/_tree.py]: Done

**Comment by mortonjt** [comment]
Alrighty, I think it should be good to go once the tests pass.

File: CHANGELOG.md [hunk 23]
```diff
@@ -3,10 +3,10 @@
 ## Version 0.4.1-dev (changes since 0.4.1 go here)
 
 ### Features
-* Added `skbio.io.format.stockholm` for reading Stockholm files into a `TabularMSA`. ([#967](https://github.com/biocore/scikit-bio/issues/967))
+* Added `skbio.tree._tree.bifurcate` for converting multifurcating trees into bifurcating trees.* Added `skbio.io.format.stockholm` for reading Stockholm files into a `TabularMSA`. ([#967](https://github.com/biocore/scikit-bio/issues/967))
```
Comment (jairideout) [inline CHANGELOG.md]: This is misformatted

Comment (jairideout) [inline CHANGELOG.md]: change `skbio.tree._tree.bifurcate` to `skbio.tree.TreeNode.bifurcate`

add https://github.com/biocore/scikit-bio/issues/896 markdown link to end of line (similar to other changelog entries)

**Comment by jairideout** [comment]
@mortonjt there are Python 3 doctest failures

File: CHANGELOG.md [hunk 24]
```diff
@@ -269,37 +269,37 @@ This is an alpha release of scikit-bio. At this stage, major backwards-incompati
 * Added ``skbio.util.cardinal_to_ordinal`` for converting a cardinal number to ordinal string (e.g., useful for error messages).
 * New I/O Registry: supports multiple file formats, automatic file format detection when reading, unified procedural ``skbio.io.read`` and ``skbio.io.write`` in addition to OOP interfaces (``read/write`` methods) on the below objects. See ``skbio.io`` for more details.
     - Added "clustal" format support:
-        * Has sniffer
-        * Readers: ``Alignment``
-        * Writers: ``Alignment``
+	 * Has sniffer
```
Comment (jairideout) [inline CHANGELOG.md]: I think your editor converted these spaces to tab characters, can you revert?

File: skbio/tree/_tree.py [hunk 25]
```diff
@@ -2716,7 +2716,64 @@ def compare_tip_distances(self, other, sample=None, dist_f=distance_from_r,
...
+                    ind = stack.pop()
+                    intermediate = TreeNode()
+                    intermediate.length = insert_length
+                    intermediate.extend(stack)
+                    n.append(intermediate)
+                    for k in stack:
+                        n.remove(k)
+                    n.extend([ind, intermediate])
+
     def index_tree(self):
```
Comment (jairideout) [inline skbio/tree/_tree.py]: You removed `@experimental(as_of="0.4.0")` from `index_tree`, can you revert?

File: skbio/tree/tests/test_tree.py [hunk 26]
```diff
@@ -817,6 +817,15 @@ def test_levelorder(self):
         obs = [n.name for n in self.simple_t.levelorder()]
         self.assertEqual(obs, exp)
 
+    def test_bifurcate(self):
```
Comment (jairideout) [inline skbio/tree/tests/test_tree.py]: Can you add a test for non-default `insert_length`? I thought I just saw tests for that...

**Comment by jairideout** [comment]
@mortonjt do you think it's worth mentioning how to remove singles using `prune` in the Notes section, now that `remove_singles` is gone?

File: skbio/tree/tests/test_tree.py (hunk 26 above)
Comment (mortonjt) [inline skbio/tree/tests/test_tree.py]: Oops.  That was accidentally deleted.  Nice catch.

**Comment by mortonjt** [comment]
The Notes section has been updated and the tests are passing.
Should be good to go!

File: CHANGELOG.md [hunk 27]
```diff
@@ -3,6 +3,7 @@
 ## Version 0.4.1-dev (changes since 0.4.1 go here)
 
 ### Features
+* Added `skbio.tree._tree.bifurcate` for converting multifurcating trees into bifurcating trees.
```
Comment (jairideout) [inline CHANGELOG.md]: @mortonjt, please see my previous comments about this changelog entry...

**Comment by jairideout** [comment]
Thanks @mortonjt, this is ready for merge once the changelog is fixed.

File: CHANGELOG.md (hunk 27 above)
Comment (mortonjt) [inline CHANGELOG.md]: I thought I addressed that issue?  Can you clarify?

Comment (jairideout) [inline CHANGELOG.md]: Please check the outdated diff

File: CHANGELOG.md (hunk 23 above)
Comment (mortonjt) [inline CHANGELOG.md]: Done

File: CHANGELOG.md [hunk 28]
```diff
@@ -3,6 +3,7 @@
 ## Version 0.4.1-dev (changes since 0.4.1 go here)
 
 ### Features
+* Added `skbio.tree.TreeNode.bifurcate` for converting multifurcating trees into bifurcating trees.
```
Comment (jairideout) [inline CHANGELOG.md]: issue number please

Comment (mortonjt) [inline CHANGELOG.md]: Sorry.  Added

File: CHANGELOG.md (hunk 23 above)
Comment (mortonjt) [inline CHANGELOG.md]: The issue number has been corrected.

File: CHANGELOG.md [hunk 29]
```diff
@@ -3,6 +3,7 @@
 ## Version 0.4.1-dev (changes since 0.4.1 go here)
 
 ### Features
+* Added `skbio.tree.TreeNode.bifurcate` for converting multifurcating trees into bifurcating trees. [#896](https://github.com/biocore/scikit-bio/issues/896))
```
Comment (jairideout) [inline CHANGELOG.md]: The issue number doesn't render correctly, can you add an opening paren?

screen shot 2016-02-17 at 11 21 32 am

Comment (mortonjt) [inline CHANGELOG.md]: :L

Fixed

**Comment by jairideout** [comment]
Thanks @mortonjt!

**Comment by mortonjt** [comment]
Thanks!!
